lambda_functions/downloader/copy_all.py

```diff
@@ -0,0 +1,145 @@
+"""Copy *all* binaries from CarbonBlack into S3, using multiprocess producer-consumer queue."""
+# This is a one-off script which backfills BinaryAlert with all CarbonBlack binaries.
+# Using dozens of processes, this will copy about 1000 binaries / minute.
+#
+# Usage: export CARBON_BLACK_URL='...' && \
+#        export ENCRYPTED_CARBON_BLACK_API_TOKEN='...' && \
+#        export TARGET_S3_BUCKET='...' && \
+#        python3 copy_all.py 2>&1 | tee copy.log
+import logging
+import multiprocessing
+import queue
+
+import cbapi
+
+if __package__:
+    from lambda_functions.downloader import main
+else:
+    import main
+
+LOG_FORMAT = '%(asctime)s %(levelname)-6s %(message)s'
+LOG_LEVEL = logging.DEBUG
+
+NUM_CONSUMERS = 32  # Number of consumer threads executing copy tasks (optimized by experiment).
+MAX_TASK_QUEUE_SIZE = NUM_CONSUMERS * 20  # Maximum number of tasks in the queue at any time.
+
+
+class CopyTask(object):
+    """A Task to copy a single binary from CarbonBlack into the BinaryAlert S3 bucket."""
+
+    def __init__(self, index, md5):
+        """Initialize a Task with the binary's information.
+
+        Args:
+            index: [int] Binary's index in the CarbonBlack enumeration.
+                Helps give the user a sense of progress.
+            md5: [string] Binary MD5, used as the key to retrieve from CarbonBlack.
+        """
+        self.index = index
+        self.md5 = md5
+
+    def __call__(self):
+        """Execute the copy task."""
+        main.download_lambda_handler({'md5': self.md5}, None)
+
+    def __str__(self):
+        """Use the index and MD5 in the string representation."""
+        return 'CopyTask [#{}] MD5:{}'.format(self.index, self.md5)
+
+
+class Consumer(multiprocessing.Process):
+    """A Consumer grabs Tasks from the shared queue and executes them asynchronously."""
+
+    def __init__(self, task_queue, failed_queue):
+        """Create a Consumer with shared communication queues.
+
+        Args:
+            task_queue: [multiprocessing.Queue] Shared queue of Tasks to perform.
+                The Consumer will process tasks until it sees a None task.
+            failed_queue: [multiprocessing.Queue] Shared queue containing the MD5 hashes of every
+                binary which failed to upload.
+        """
+        super(Consumer, self).__init__()
+        self.task_queue = task_queue
+        self.failed_queue = failed_queue
+
+        # Each process needs its own logger to avoid race conditions.
+        self.logger = logging.getLogger(self.name)
```

Comment:
As an aside, the logging package actually handles race conditions pretty well through the use of locks, etc (as seen [here](https://github.com/python/cpython/blob/master/Lib/logging/__init__.py#L201). That said, this is dependent on the `threading` package being available, and I don't think you'll have that in lambda(?). So, in normal use cases, having a separate logger for each is probably unnecessary, but is probably the safer/better solution in this case.

Reply:
This is actually a one-off script that is `not` designed to run in Lambda. It's probably confusing that it's part of the `lambda_functions` hierarchy, but I'm not sure where would be a better place to put it

Reply:
You're right; [Python logging is thread-safe](https://docs.python.org/3/library/logging.html#thread-safety) I'll remove this